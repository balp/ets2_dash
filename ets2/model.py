import json
from dataclasses import dataclass
from typing import Optional, Dict, List

from ets2.config import TruckConfig, truckconfig_from_dict, TrailerConfig, trailer_config_from_dict
from ets2.jobs import JobConfig, jobconfig_from_dict
from ets2.telematic import Telematic, telematic_from_dict
from ets2.tracks import Tracks


@dataclass
class Info:
    paused: bool


def info_from_dict(data: Dict) -> Info:
    return Info(paused=data['paused'])


@dataclass
class Version:
    major: int
    minor: int


def version_from_dict(data: Dict) -> Version:
    return Version(data['major'], data['minor'])


@dataclass
class Game:
    id: str
    name: str
    raw_version: int
    version: Version


def game_from_dict(data: Dict) -> Game:
    return Game(id=data['id'],
                name=data['name'],
                raw_version=data['raw_version'],
                version=version_from_dict(data['version']))


class Model:
    def __init__(self):
        self._observers = []
        self.telematic: Optional[Telematic] = None
        self.job: Optional[JobConfig] = None
        self.info: Optional[Info] = None
        self.game: Optional[Game] = None
        self.truck_config: Optional[TruckConfig] = None
        self.trailer_config: List[Optional[TrailerConfig]] = [None for _ in range(10)]
        self.tracks: Tracks = Tracks()

    def register_observer(self, observer):
        self._observers.append(observer)

    def _data_changed(self, change):
        for observer in self._observers:
            observer.notify(self, change)

    def set_telematic_data(self, data):
        self.telematic = telematic_from_dict(data)
        self.tracks.add_telematic(self.telematic)
        self._data_changed("telematic")

    def set_job_config(self, data):
        self.job = jobconfig_from_dict(data)
        self._data_changed("job")

    def set_info(self, data):
        self.info = info_from_dict(data)
        self._data_changed("info")

    def set_game(self, data):
        self.game = game_from_dict(data)
        self._data_changed("game")

    def set_truck_config(self, data):
        self.truck_config = truckconfig_from_dict(data)
        self._data_changed("truck_config")

    def set_trailer_config(self, data, index):
        self.trailer_config[index] = trailer_config_from_dict(data)
        self._data_changed("trailer_config")

    def get_game_pause(self) -> Optional[bool]:
        if self.info:
            return self.info.paused
        return None

    def get_game_name(self) -> Optional[str]:
        if self.game:
            return self.game.name
        return None

    def get_game_id(self) -> Optional[str]:
        if self.game:
            return self.game.id
        return None

    def get_time_left(self) -> Optional[int]:
        if self.telematic and self.job:
            game_time = self.telematic.common.game_time
            delivery_time: int = self.job.delivery_time
            if game_time and delivery_time and delivery_time != 0xFFFFFFFF:
                return delivery_time - game_time
        return None

    def get_time_to_rest(self) -> int:
        if self.telematic:
            return self.telematic.common.rest_stop
        return 0

    def get_time_destination(self) -> int:
        if self.telematic:
            eta = self.telematic.truck.navigation_time
            return int(eta) // 60
        return 0

    def get_time_destination_with_rest(self) -> int:
        if self.telematic:

            drive_time = self.get_time_destination()
            next_rest = self.get_time_to_rest()
            total_time = drive_time

            if drive_time > next_rest:
                total_time += 9*60
                total_time += ((drive_time - next_rest) // (11 * 60)) * 9 * 60
            return int(total_time)
        return 0

    def get_speed_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.speed * 3.6
        return 0.0

    def get_cruise_control_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.cruise_control * 3.6
        return 0.0

    def get_speed_limit_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.navigation_speed_limit * 3.6
        return 0.0

    def get_speed_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.speed * 2.2369363
        return 0.0

    def get_cruise_control_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.cruise_control * 2.2369363
        return 0.0

    def get_speed_limit_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.navigation_speed_limit * 2.2369363
        return 0.0

    def get_fuel_left(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_amount
        return 0.0

    def get_fuel_range(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_range
        return 0.0

    def get_fuel_consumtion(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_consumption_average
        return 0.0

    def get_wear_cabin(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_cabin
        return 0.0

    def get_wear_chassis(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_chassis
        return 0.0

    def get_wear_engine(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_engine
        return 0.0

    def get_wear_transmission(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_transmission
        return 0.0

    def get_wear_wheels(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_wheels
        return 0.0

    def get_wear_trailer(self) -> float:
        if self.telematic:
            return self.telematic.trailer.wear_chassis
        return 0.0

    def get_light_high_beam(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beam_high
        return False

    def get_light_low_beam(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beam_low
        return False

    def get_light_beacon(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beacon
        return False

    def get_light_l_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_lblinker
        return False

    def get_light_r_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_rblinker
        return False

    def get_l_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.lblinker
        return False

    def get_r_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.rblinker
        return False

    def get_light_parking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_parking
        return False

    def get_light_reverse(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_reverse
        return False

    def get_light_aux_front(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_aux_front != 0
        return False

    def get_light_aux_roof(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_aux_roof != 0
        return False

    def get_light_breaking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_brake
        return False

    def get_ad_blue_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.adblue_warning
        return False

    def get_break_emergency(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure_emergency
        return False

    def get_break_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure_warning
        return False

    def get_break_motor(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_motor
        return False

    def get_break_parking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_parking
        return False

    def get_electric(self) -> bool:
        if self.telematic:
            return self.telematic.truck.electric_enabled
        return False

    def get_battery_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.battery_voltage_warning
        return False

    def get_engine(self) -> bool:
        if self.telematic:
            return self.telematic.truck.engine_enabled
        return False

    def get_fuel_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.fuel_warning
        return False

    def get_oil_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.oil_pressure_warning
        return False

    def get_water_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.water_temperature_warning
        return False

    def get_wipers(self) -> bool:
        if self.telematic:
            return self.telematic.truck.wipers
        return False

    def get_air_pressure(self) -> float:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure
        return False

    def get_break_retarder(self) -> int:
        if self.telematic:
            return self.telematic.truck.brake_retarder
        return False

    def get_break_temperature(self) -> float:
        if self.telematic:
            return self.telematic.truck.brake_temperature
        return False


def add_json_to_model(model: Model, json_data: json, topic: str):
    if topic == "ets2/data":
        model.set_telematic_data(json_data)
    elif topic == "ets2/game":
        model.set_game(json_data)
    elif topic == "ets2/info":
        model.set_info(json_data)
    elif topic == "ets2/info/config/job":
        model.set_job_config(json_data)
    elif topic == "ets2/info/config/truck":
        model.set_truck_config(json_data)
    elif topic == "ets2/info/config/trailer.0":
        model.set_trailer_config(json_data, 0)
    elif topic == "ets2/info/config/trailer.1":
        model.set_trailer_config(json_data, 1)
    elif topic == "ets2/info/config/trailer.2":
        model.set_trailer_config(json_data, 2)
    elif topic == "ets2/info/config/trailer.3":
        model.set_trailer_config(json_data, 3)
    elif topic == "ets2/info/config/trailer.4":
        model.set_trailer_config(json_data, 4)
    elif topic == "ets2/info/config/trailer.5":
        model.set_trailer_config(json_data, 5)
    elif topic == "ets2/info/config/trailer.6":
        model.set_trailer_config(json_data, 6)
    elif topic == "ets2/info/config/trailer.7":
        model.set_trailer_config(json_data, 7)
    elif topic == "ets2/info/config/trailer.8":
        model.set_trailer_config(json_data, 8)
    elif topic == "ets2/info/config/trailer.9":
        model.set_trailer_config(json_data, 9)
