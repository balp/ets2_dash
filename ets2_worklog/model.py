import datetime
import json
from dataclasses import dataclass
from typing import Optional, Dict, List

import ets2.model


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class Job:
    config: Optional[ets2.model.JobConfig]
    started: Optional[datetime.datetime]
    ended: Optional[datetime.datetime]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.model.Tracks


class WorkLog:
    """A log work jobs in ETS2/ATS"""

    def __init__(self, data: ets2.model.Model):
        self._model = data
        self._model.register_observer(self)
        self._jobs: List[Job] = []

    def __repr__(self):
        return str(self._jobs)

    def notify(self, model: ets2.model.Model, change: str):
        if self._jobs:
            if model.job != self._jobs[-1].config:
                self._jobs[-1].ended = self.time_in_game(model)
                job = self.job_from_model(model)
                self._jobs.append(job)
            else:
                if model.telematic:
                    self._jobs[-1].track.add_telematic(model.telematic)
        else:
            job = self.job_from_model(model)
            self._jobs.append(job)

    def job_from_model(self, model):
        started = self.time_in_game(model)
        job = Job(config=model.job,
                  started=started,
                  ended=None,
                  delivered=None,
                  cancelled=None,
                  track=ets2.model.Tracks())
        return job

    def time_in_game(self, model):
        if model.telematic:
            started = datetime.datetime.fromordinal(1) \
                      + datetime.timedelta(minutes=model.telematic.common.game_time)
        else:
            started = None
        return started

    def job_delivered(self, delivered: Delivered) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].delivered = delivered

    def job_cancelled(self, cancelled: Cancelled) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].cancelled = cancelled


def add_json_to_work_log(work_log: WorkLog, json_data: json, topic: str) -> None:
    if topic == "ets2/info/gameplay/job.cancelled":
        work_log.job_cancelled(cancelled_from_dict(json_data))
    elif topic == "ets2/info/gameplay/job.delivered":
        work_log.job_delivered(delivered_from_dict(json_data))
