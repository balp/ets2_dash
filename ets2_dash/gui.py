
import paho.mqtt.client as mqtt
import PySimpleGUI
import json
import typing
import logging
import threading
from dataclasses import dataclass

from ets2_dash.model import Model
from ets2_dash.view import View

@dataclass
class GlobalState:
    active: bool = True


def on_message(client: typing.Any, userdata: typing.Any , message : object):
    # print("message received ", message.payload)
    # print("message topic=",message.topic)
    assert isinstance(userdata, Model)
    model = userdata
    json_data = json.loads(message.payload)
    if message.topic == "ets2/data":
        model.set_telematic_data(json_data)
    elif message.topic == "ets2/game":
        model.set_game(json_data)
    elif message.topic == "ets2/info":
        model.set_info(json_data)
    elif message.topic == "ets2/info/config/job":
        model.set_job_config(json_data)
    elif message.topic == "ets2/info/config/truck":
        model.set_truck_config(json_data)
    elif message.topic == "ets2/info/config/trailer.0":
        model.set_trailer_config(json_data, 0)
    elif message.topic == "ets2/info/config/trailer.1":
        model.set_trailer_config(json_data, 1)
    elif message.topic == "ets2/info/config/trailer.2":
        model.set_trailer_config(json_data, 2)
    elif message.topic == "ets2/info/config/trailer.3":
        model.set_trailer_config(json_data, 3)
    elif message.topic == "ets2/info/config/trailer.4":
        model.set_trailer_config(json_data, 4)
    elif message.topic == "ets2/info/config/trailer.5":
        model.set_trailer_config(json_data, 5)
    elif message.topic == "ets2/info/config/trailer.6":
        model.set_trailer_config(json_data, 6)
    elif message.topic == "ets2/info/config/trailer.7":
        model.set_trailer_config(json_data, 7)
    elif message.topic == "ets2/info/config/trailer.8":
        model.set_trailer_config(json_data, 8)
    elif message.topic == "ets2/info/config/trailer.9":
        model.set_trailer_config(json_data, 9)


def mqtt_thread_loop(model: Model, state: GlobalState):
    client = mqtt.Client("ets2_gui", userdata=model) #create new instance
    client.on_message = on_message
    client.connect("localhost") #connect to broker
    client.subscribe("ets2/data")
    client.subscribe("ets2/game")
    client.subscribe("ets2/info")
    client.subscribe("ets2/info/config/job")
    client.subscribe("ets2/info/config/substances")
    # client.subscribe("ets2/info/config/controls")
    # client.subscribe("ets2/info/config/hshifter")
    client.subscribe("ets2/info/config/truck")
    # client.subscribe("ets2/info/config/trailer")

    client.subscribe("ets2/info/config/trailer.0")
    client.subscribe("ets2/info/config/trailer.1")
    client.subscribe("ets2/info/config/trailer.2")
    client.subscribe("ets2/info/config/trailer.3")
    client.subscribe("ets2/info/config/trailer.4")
    client.subscribe("ets2/info/config/trailer.5")
    client.subscribe("ets2/info/config/trailer.6")
    client.subscribe("ets2/info/config/trailer.7")
    client.subscribe("ets2/info/config/trailer.8")
    client.subscribe("ets2/info/config/trailer.9")

    while state.active:
        client.loop(timeout=1.0)

def main():
    print("Startup!!!")
    model: Model = Model()
    state: GlobalState = GlobalState()
    mqtt_reader_thread = threading.Thread(target=mqtt_thread_loop, args=(model, state))
    mqtt_reader_thread.start()

    PySimpleGUI.ChangeLookAndFeel('Dark')
    hmi = View(model)

    print("Loop!!!")
    while (True):
        event, values = hmi.window.Read(timeout=50)
        hmi.update_data()
        if event == 'Exit':
            break
    state.active = False
    mqtt_reader_thread.join(timeout=2.0)
    hmi.window.CloseNonBlockingForm()


if __name__ == '__main__':
    main()