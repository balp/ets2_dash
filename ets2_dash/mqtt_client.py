import paho.mqtt.client as mqtt
import typing

def on_message(client : typing.Any, userdata : typing.Any, message : object):
    print("message received " , message.payload)
    print("message topic=",message.topic)
    print("message qos=",message.qos)
    print("message retain flag=",message.retain)

def main():
    client = mqtt.Client("ets2_dash") #create new instance
    client.on_message = on_message
    client.connect("localhost") #connect to broker
    client.subscribe("ets2/data")
    client.subscribe("ets2/info/config/substances")
    client.subscribe("ets2/info/config/job")
    client.subscribe("ets2/info/config/controls")
    client.subscribe("ets2/info/config/hshifter")
    client.subscribe("ets2/info/config/truck")
    client.subscribe("ets2/info/config/trailer")

    client.loop_forever()

if __name__ == '__main__':
    main()