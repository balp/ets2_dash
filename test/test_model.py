import json
import pprint

from approvaltests import verify

from ets2_dash.model import Model
import pytest


def test_get_time_left_normal():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    model.set_job_config(json.loads("{}"))
    assert model.get_time_left() is None


def test_get_time_left_community():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    with open("data/job_community.json", "r") as j:
        model.set_job_config(json.load(j))
    assert model.get_time_left() is None


def test_get_time_to_rest():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_time_to_rest() == 63


def test_get_time_destination_freeride():
    model = Model()
    with open("data/telematic_freeride.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_time_destination() == 100


def test_get_time_destination():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_time_destination() == 0


def test_get_time_destination_with_rest_no_rest():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_time_destination_with_rest() == 0


def test_get_time_destination_with_rest_needs_rest():
    model = Model()
    with open("data/telematic_reststop.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert 1263 == model.get_time_destination_with_rest()


def test_get_speed_kmh():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_speed_kmh() == pytest.approx(16.1, abs=0.1)


def test_get_cruise_control_kmh():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0.000, abs=0.001) == model.get_cruise_control_kmh()


def test_get_speed_limit_kmh():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_speed_limit_kmh() == pytest.approx(0.0, abs=0.1)


def test_get_speed_mph():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_speed_mph() == pytest.approx(10.0, abs=0.1)


def test_get_cruise_control_mph():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0.000, abs=0.001) == model.get_cruise_control_mph()


def test_get_speed_limit_mph():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_speed_limit_mph() == pytest.approx(0.0, abs=0.1)


def test_get_fuel_left():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_fuel_left() == pytest.approx(142.3, abs=0.05)


def test_get_fuel_range():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_fuel_range() == pytest.approx(444.6, abs=0.05)


def test_get_fuel_consumtion():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_fuel_consumtion() == pytest.approx(0.5, abs=0.1)


def test_get_wear_cabin():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0.005, abs=0.005) == model.get_wear_cabin()


def test_get_wear_chassis():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0.006, abs=0.005) == model.get_wear_chassis()


def test_get_wear_engine():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0.004, abs=0.005) == model.get_wear_engine()


def test_get_wear_transmission():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0.003, abs=0.005) == model.get_wear_transmission()


def test_get_wear_wheels():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_wear_wheels() == pytest.approx(0.005, abs=0.001)


def test_get_wear_trailer():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0.00, abs=0.005) == model.get_wear_trailer()


def test_get_game_pause():
    model = Model()
    with open("data/info.json", "r") as j:
        model.set_info(json.load(j))
    assert not model.get_game_pause()


def test_get_game_name():
    model = Model()
    with open("data/game.json", "r") as j:
        model.set_game(json.load(j))
    assert "Euro Truck Simulator 2 1.33.2.19s" == model.get_game_name()


def test_get_game_id():
    model = Model()
    with open("data/game.json", "r") as j:
        model.set_game(json.load(j))
    assert "eut2" == model.get_game_id()


def test_get_light_high_beam():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_light_high_beam()


def test_get_light_low_beam():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_light_low_beam()


def test_get_light_beacon():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_light_beacon()


def test_get_light_l_blinker():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_light_l_blinker()


def test_get_light_r_blinker():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_light_r_blinker()


def test_get_light_parking():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_light_parking()


def test_get_light_reverse():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_light_reverse()


def test_get_light_aux_front():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_light_aux_front()


def test_get_light_aux_roof():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_light_aux_roof()


def test_get_light_breaking():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_light_breaking()


def test_get_l_blinker():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_l_blinker()


def test_get_r_blinker():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_r_blinker()


def test_get_ad_blue_warning():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_ad_blue_warning()


def test_get_break_motor_breaking():
    model = Model()
    with open("data/telematic_breaking.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_break_motor()


def test_get_break_motor_breaking_engine():
    model = Model()
    with open("data/telematic_engine_break.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_break_motor()


def test_get_break_motor():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_break_motor()


def test_get_break_parking():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_break_parking()


def test_get_electric():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_electric()


def test_get_battery_warning():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_battery_warning()


def test_get_engine():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_engine()


def test_get_fuel_warning():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_fuel_warning()


def test_get_oil_warning():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_oil_warning()


def test_get_water_warning():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_water_warning()


def test_get_wipers():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_wipers()


def test_get_break_emergency():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_break_emergency()


def test_get_break_warning():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_break_warning()


def test_get_air_pressure():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_air_pressure() == pytest.approx(115.5, abs=0.5)


def test_get_break_retarder():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert pytest.approx(0, abs=0.5) == model.get_break_retarder()


def test_get_break_temperature():
    model = Model()
    with open("data/telematic.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert model.get_break_temperature() == pytest.approx(27.8, abs=0.5)


def test_get_break_warning_1_01():
    model = Model()
    with open("data/telematic_1_01.json", "r") as j:
        model.set_telematic_data(json.load(j))
    assert not model.get_break_warning()


def test_dual_trailer_config():
    model = Model()
    with open("data/trailer_dual_0.json", "r") as t0:
        model.set_trailer_config(json.load(t0), 0)
    with open("data/trailer_dual_1.json", "r") as t1:
        model.set_trailer_config(json.load(t1), 1)
    assert model.trailer_config[0] is not None
    assert model.trailer_config[1] is not None
    assert model.trailer_config[2] is None


def test_no_trailer():
    model = Model()
    data = json.loads("{}")
    pprint.pprint(data)
    model.set_trailer_config(data, 0)
    model.set_trailer_config(data, 1)
    model.set_trailer_config(data, 1)
    assert model.trailer_config[0] is None
    assert model.trailer_config[1] is None
    assert model.trailer_config[2] is None


def test_add_job():
    model = Model()
    with open("data/job_1.01.json", "r") as j:
        model.set_job_config(json.load(j))
    assert model.job.income == 24293


def test_add_no_job():
    model = Model()
    model.set_job_config(json.loads("{}"))
    assert model.job is None


def test_inital_track():
    model = Model()
    files = ["data/telematic.json",
             "data/telematic_1_01.json",
             "data/telematic_breaking.json",
             "data/telematic_engine_break.json",
             "data/telematic_freeride.json",
             "data/telematic_reststop.json"]
    for f in files:
        with open(f, "r") as j:
            model.set_telematic_data(json.load(j))
    verify(str(model.tracks))


def test_inital_track_bottom_left_corner_empty():
    model = Model()
    assert model.tracks.bottom_left() == (-100000, -5)

def test_inital_track_bottom_left_corner():
    model = Model()
    files = ["data/telematic.json",
             "data/telematic_1_01.json",
             "data/telematic_breaking.json",
             "data/telematic_engine_break.json",
             "data/telematic_freeride.json",
             "data/telematic_reststop.json"]
    for f in files:
        with open(f, "r") as j:
            model.set_telematic_data(json.load(j))
    assert model.tracks.bottom_left() == (-38626.81513977051, -1.187657356262207)

def test_inital_track_top_right_corner_empty():
    model = Model()
    assert model.tracks.top_right() == (61000, 30)


def test_inital_track_top_right_corner():
    model = Model()
    files = ["data/telematic.json",
             "data/telematic_1_01.json",
             "data/telematic_breaking.json",
             "data/telematic_engine_break.json",
             "data/telematic_freeride.json",
             "data/telematic_reststop.json"]
    for f in files:
        with open(f, "r") as j:
            model.set_telematic_data(json.load(j))
    assert model.tracks.top_right() == (60740.99865722656, 28.410934448242188)
