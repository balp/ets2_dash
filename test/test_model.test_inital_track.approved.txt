[TrackPoint(position=Vector(x=-38626.81513977051, y=28.410934448242188, z=35655.31903076172), time=296378)]